services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/validation/Validators.kt

```diff
@@ -11,67 +9,104 @@ private val ValidationSuccess: ValidationError? = null
  * Keeps track of the paths of validation failures, but not the errors that occurred.
  *
  * @see HankePublicValidator for examples on usage.
+ * @see ApplicationValidator for examples on usage.
  */
 object Validators {
 
     /**
      * Starts a validation chain. Further validations can be chained with [ValidationResult.and].
      */
-    fun validate(f: () -> ValidationError?): ValidationResult {
-        return ValidationResult().and(f)
+    fun validate(f: () -> ValidationResult): ValidationResult {
+        return ValidationResult.success().and(f)
     }
 
-    fun validateFalse(condition: Boolean, path: String): ValidationError? =
-        if (condition) path else ValidationSuccess
+    fun validateTrue(condition: Boolean, path: String): ValidationResult =
+        if (condition) ValidationResult.success() else ValidationResult.failure(path)
+
+    fun validateFalse(condition: Boolean, path: String): ValidationResult =
+        if (!condition) ValidationResult.success() else ValidationResult.failure(path)
 
-    fun validateFalse(condition: Boolean?, path: String): ValidationError? =
-        if (condition == false) ValidationSuccess else path
+    fun validateFalse(condition: Boolean?, path: String): ValidationResult =
+        if (condition == false) ValidationResult.success() else ValidationResult.failure(path)
 
-    fun notNull(value: Any?, path: String): ValidationError? = validateFalse(value == null, path)
+    fun notNull(value: Any?, path: String): ValidationResult = validateFalse(value == null, path)
 
-    fun notBlank(value: String, path: String): ValidationError? =
+    fun notBlank(value: String, path: String): ValidationResult =
         validateFalse(value.isBlank(), path)
 
-    fun notNullOrBlank(value: String?, path: String): ValidationError? =
+    /**
+     * Fails if the string is defined, non-empty and has nothing but whitespace characters.
+     * - null -> Ok
+     * - "" -> Ok
+     * - " " -> Fail
+     * - "*" -> Ok
+     * - "\t\n \t\n" -> Fail
+     * - " f " -> Ok
+     */
+    fun notJustWhitespace(value: String?, path: String): ValidationResult =
```

Comment:
How about notBlank?

Reply:
There's already a `notBlank()`, which calls `String.isBlank()`. That one considers both null and empty strings as blanks.

This was not easy to name, because it's different than the traditional definition of a blank string. I couldn't find a perfect name, so I went with one that's hopefully clear after reading the associated doc, but unclear enough to maybe get the reader to check the doc.